File: examples/hello-world/gabriel-greeting-node/_internal/server_test.js
'use strict';

const test = require('node:test');
const assert = require('node:assert/strict');
const grpc = require('@grpc/grpc-js');
const { observability } = require('@organic-programming/holons');

const pb = require('../gen/node/greeting/v1/greeting_pb.js');
const grpcPb = require('../gen/node/greeting/v1/greeting_grpc_pb.js');
const { GreetingService, listenAndServe } = require('./server');

function startServer() {
  const server = new grpc.Server();
  server.addService(grpcPb.GreetingServiceService, new GreetingService());

  return new Promise((resolve, reject) => {
    server.bindAsync('127.0.0.1:0', grpc.ServerCredentials.createInsecure(), (error, port) => {
      if (error) {
        reject(error);
        return;
      }
      server.start();
      resolve({
        server,
        client: new grpcPb.GreetingServiceClient(`127.0.0.1:${port}`, grpc.credentials.createInsecure()),
      });
    });
  });
}

function callUnary(client, method, request) {
  return new Promise((resolve, reject) => {
    client[method](request, (error, response) => {
      if (error) {
        reject(error);
        return;
      }
      resolve(response);
    });
  });
}

test('RPC ListLanguages returns all languages', async (t) => {
  const runtime = await startServer();
  t.after(() => runtime.client.close());
  t.after(() => runtime.server.forceShutdown());

  const response = await callUnary(runtime.client, 'listLanguages', new pb.ListLanguagesRequest());

  assert.equal(response.getLanguagesList().length, 56);
});

test('RPC ListLanguages populates required fields', async (t) => {
  const runtime = await startServer();
  t.after(() => runtime.client.close());
  t.after(() => runtime.server.forceShutdown());

  const response = await callUnary(runtime.client, 'listLanguages', new pb.ListLanguagesRequest());

  for (const language of response.getLanguagesList()) {
    assert.ok(language.getCode());
    assert.ok(language.getName());
    assert.ok(language.getNative());
  }
});

test('RPC SayHello uses requested language', async (t) => {
  const runtime = await startServer();
  t.after(() => runtime.client.close());
  t.after(() => runtime.server.forceShutdown());

  const request = new pb.SayHelloRequest();
  request.setName('Bob');
  request.setLangCode('fr');

  const response = await callUnary(runtime.client, 'sayHello', request);

  assert.equal(response.getGreeting(), 'Bonjour Bob');
  assert.equal(response.getLanguage(), 'French');
  assert.equal(response.getLangCode(), 'fr');
});

test('RPC SayHello uses localized default name', async (t) => {
  const runtime = await startServer();
  t.after(() => runtime.client.close());
  t.after(() => runtime.server.forceShutdown());

  const request = new pb.SayHelloRequest();
  request.setLangCode('fr');

  const response = await callUnary(runtime.client, 'sayHello', request);

  assert.equal(response.getGreeting(), 'Bonjour Marie');
  assert.equal(response.getLangCode(), 'fr');
});

test('RPC SayHello falls back to English', async (t) => {
  const runtime = await startServer();
  t.after(() => runtime.client.close());
  t.after(() => runtime.server.forceShutdown());

  const request = new pb.SayHelloRequest();
  request.setName('Bob');
  request.setLangCode('xx');

  const response = await callUnary(runtime.client, 'sayHello', request);

  assert.equal(response.getGreeting(), 'Hello Bob');
  assert.equal(response.getLangCode(), 'en');
});

test('SayHello emits observability signals with stdio transport', async (t) => {
  const oldOpObs = process.env.OP_OBS;
  process.env.OP_OBS = 'logs,metrics';
  observability.reset();
  const obs = observability.configure({ slug: 'gabriel-greeting-node' });
  t.after(() => {
    observability.reset();
    if (oldOpObs === undefined) {
      delete process.env.OP_OBS;
    } else {
      process.env.OP_OBS = oldOpObs;
    }
  });

  const server = await listenAndServe('stdio://');
  t.after(async () => {
    if (server.__holonsRuntime) await server.stopHolon();
  });

  const request = new pb.SayHelloRequest();
  request.setName(' Bob ');
  request.setLangCode('en');

  const response = await new Promise((resolve, reject) => {
    new GreetingService().sayHello({ request }, (error, out) => {
      if (error) {
        reject(error);
        return;
      }
      resolve(out);
    });
  });

  assert.equal(response.getGreeting(), 'Hello Bob');
  const snapshot = obs.registry.snapshot();
  const counter = snapshot.counters.find((sample) => sample.name === 'greeting_emitted_total');
  assert.ok(counter);
  assert.deepEqual(counter.labels, {
    lang_code: 'en',
    language: 'English',
    transport: 'stdio',
  });
  assert.equal(counter.value, 1);

  const entry = obs.logRing.drain().find((logEntry) => observability.bodyString(logEntry) === 'Greeted Bob in English (en)');
  assert.ok(entry);
  const attrs = ['duration_ns', 'greeting', 'lang_code', 'language', 'name', 'transport']
    .map((key) => entry.attributes.find((attr) => attr.key === key));
  assert.equal(attrs.every(Boolean), true);
  assert.deepEqual(attrs.map((attr) => attr.key).sort(), [
    'duration_ns',
    'greeting',
    'lang_code',
    'language',
    'name',
    'transport',
  ]);
  assert.deepEqual(attr(entry, 'holons.slug').value, { string_value: 'gabriel-greeting-node' });
  assert.deepEqual(attr(entry, 'service.name').value, { string_value: 'gabriel-greeting-node' });
  assert.ok(attr(entry, 'holons.session_id').value.string_value);
  assert.deepEqual(attr(entry, 'lang_code').value, { string_value: 'en' });
  assert.deepEqual(attr(entry, 'language').value, { string_value: 'English' });
  assert.deepEqual(attr(entry, 'name').value, { string_value: 'Bob' });
  assert.deepEqual(attr(entry, 'greeting').value, { string_value: 'Hello Bob' });
  assert.deepEqual(attr(entry, 'transport').value, { string_value: 'stdio' });
  assert.equal(Object.prototype.hasOwnProperty.call(attr(entry, 'duration_ns').value, 'int_value'), true);
  assert.ok(Number(attr(entry, 'duration_ns').value.int_value) >= 0);
});

function attr(record, key) {
  const found = (record.attributes || []).find((item) => item.key === key);
  assert.ok(found, `missing attribute ${key}`);
  return found;
}

File: examples/hello-world/gabriel-greeting-dart/_internal/server_test.dart
import 'dart:io';

import 'package:grpc/grpc.dart';
import 'package:holons/gen/holons/v1/observability.pb.dart' as obspb;
import 'package:holons/holons.dart' as holons;
import 'package:test/test.dart';

import '../gen/dart/greeting/v1/greeting.pbgrpc.dart';
import 'server.dart';

void main() {
  test('RPC ListLanguages returns all languages', () async {
    final server = Server.create(services: <Service>[GreetingService()]);
    await server.serve(address: InternetAddress.loopbackIPv4, port: 0);
    addTearDown(server.shutdown);

    final channel = ClientChannel(
      '127.0.0.1',
      port: server.port!,
      options: const ChannelOptions(credentials: ChannelCredentials.insecure()),
    );
    addTearDown(channel.shutdown);

    final client = GreetingServiceClient(channel);
    final response = await client.listLanguages(ListLanguagesRequest());

    expect(response.languages, hasLength(56));
  });

  test('RPC ListLanguages populates required fields', () async {
    final server = Server.create(services: <Service>[GreetingService()]);
    await server.serve(address: InternetAddress.loopbackIPv4, port: 0);
    addTearDown(server.shutdown);

    final channel = ClientChannel(
      '127.0.0.1',
      port: server.port!,
      options: const ChannelOptions(credentials: ChannelCredentials.insecure()),
    );
    addTearDown(channel.shutdown);

    final client = GreetingServiceClient(channel);
    final response = await client.listLanguages(ListLanguagesRequest());

    for (final language in response.languages) {
      expect(language.code, isNotEmpty);
      expect(language.name, isNotEmpty);
      expect(language.native, isNotEmpty);
    }
  });

  test('RPC SayHello uses requested language', () async {
    final server = Server.create(services: <Service>[GreetingService()]);
    await server.serve(address: InternetAddress.loopbackIPv4, port: 0);
    addTearDown(server.shutdown);

    final channel = ClientChannel(
      '127.0.0.1',
      port: server.port!,
      options: const ChannelOptions(credentials: ChannelCredentials.insecure()),
    );
    addTearDown(channel.shutdown);

    final client = GreetingServiceClient(channel);
    final response = await client.sayHello(
      SayHelloRequest()
        ..name = 'Bob'
        ..langCode = 'fr',
    );

    expect(response.greeting, equals('Bonjour Bob'));
    expect(response.language, equals('French'));
    expect(response.langCode, equals('fr'));
  });

  test('RPC SayHello uses localized default name', () async {
    final server = Server.create(services: <Service>[GreetingService()]);
    await server.serve(address: InternetAddress.loopbackIPv4, port: 0);
    addTearDown(server.shutdown);

    final channel = ClientChannel(
      '127.0.0.1',
      port: server.port!,
      options: const ChannelOptions(credentials: ChannelCredentials.insecure()),
    );
    addTearDown(channel.shutdown);

    final client = GreetingServiceClient(channel);
    final response = await client.sayHello(
      SayHelloRequest()..langCode = 'fr',
    );

    expect(response.greeting, equals('Bonjour Marie'));
    expect(response.langCode, equals('fr'));
  });

  test('RPC SayHello falls back to English', () async {
    final server = Server.create(services: <Service>[GreetingService()]);
    await server.serve(address: InternetAddress.loopbackIPv4, port: 0);
    addTearDown(server.shutdown);

    final channel = ClientChannel(
      '127.0.0.1',
      port: server.port!,
      options: const ChannelOptions(credentials: ChannelCredentials.insecure()),
    );
    addTearDown(channel.shutdown);

    final client = GreetingServiceClient(channel);
    final response = await client.sayHello(
      SayHelloRequest()
        ..name = 'Bob'
        ..langCode = 'xx',
    );

    expect(response.greeting, equals('Hello Bob'));
    expect(response.langCode, equals('en'));
  });

  test('RPC SayHello emits greeting observability at the handler boundary',
      () async {
    holons.reset();
    final obs = holons.fromEnv(
      const holons.Config(
        slug: 'gabriel-greeting-dart-test',
        instanceUid: 'greeting-test-uid',
      ),
      const {'OP_OBS': 'logs,metrics'},
    );
    holons.setCurrentTransport('stdio');
    addTearDown(() {
      holons.setCurrentTransport('');
      holons.reset();
    });

    final server = Server.create(services: <Service>[GreetingService()]);
    await server.serve(address: InternetAddress.loopbackIPv4, port: 0);
    addTearDown(server.shutdown);

    final channel = ClientChannel(
      '127.0.0.1',
      port: server.port!,
      options: const ChannelOptions(credentials: ChannelCredentials.insecure()),
    );
    addTearDown(channel.shutdown);

    final client = GreetingServiceClient(channel);
    final response = await client.sayHello(
      SayHelloRequest()
        ..name = 'Bob'
        ..langCode = 'fr',
    );

    expect(response.greeting, equals('Bonjour Bob'));

    final entry = obs.logRing!.drain().singleWhere(
          (log) => log.message == 'Greeted Bob in French (fr)',
        );
    final record = holons.toProtoLogRecord(entry);
    final attrs = _attrs(record.attributes);
    expect(record.body.stringValue, equals('Greeted Bob in French (fr)'));
    expect(record.severityNumber,
        equals(obspb.SeverityNumber.SEVERITY_NUMBER_INFO));
    expect(attrs[holons.attrHolonsSlug]?.stringValue,
        equals('gabriel-greeting-dart-test'));
    expect(attrs[holons.attrServiceName]?.stringValue,
        equals('gabriel-greeting-dart-test'));
    expect(attrs[holons.attrHolonsInstanceUid]?.stringValue,
        equals('greeting-test-uid'));
    expect(attrs[holons.attrServiceInstanceId]?.stringValue,
        equals('greeting-test-uid'));
    expect(attrs[holons.attrHolonsSessionId]?.stringValue, equals(''));
    expect(
      attrs.keys.where(
          (key) => !key.startsWith('holons.') && !key.startsWith('service.')),
      unorderedEquals(<String>[
        'lang_code',
        'language',
        'name',
        'greeting',
        'transport',
        'duration_ns',
        holons.attrLoggerName,
        holons.attrCodeCaller,
      ]),
    );
    expect(attrs['lang_code']?.stringValue, equals('fr'));
    expect(attrs['language']?.stringValue, equals('French'));
    expect(attrs['name']?.stringValue, equals('Bob'));
    expect(attrs['greeting']?.stringValue, equals('Bonjour Bob'));
    expect(attrs['transport']?.stringValue, equals('stdio'));
    expect(attrs['duration_ns']?.whichValue(),
        equals(obspb.AnyValue_Value.intValue));
    expect(attrs['duration_ns']!.intValue.toInt(), greaterThanOrEqualTo(0));

    final counter = obs.registry!.listCounters().singleWhere(
          (metric) => metric.name == 'greeting_emitted_total',
        );
    expect(counter.value(), equals(1));
    expect(
      counter.labels,
      equals(<String, String>{
        'lang_code': 'fr',
        'language': 'French',
        'transport': 'stdio',
      }),
    );
  });
}

Map<String, obspb.AnyValue> _attrs(Iterable<obspb.KeyValue> attrs) => {
      for (final attr in attrs) attr.key: attr.value,
    };
